beeline-params/src/Beeline/Params/HeaderSchema.hs

```diff
@@ -0,0 +1,198 @@
+      in
+        if BS.null cookiesBS
+          then mempty
+          else DList.singleton (HTTPTypes.hCookie, cookiesBS)
+
+encodeHeader :: PD.ParameterDefinition param -> param -> HeaderBuilder
+encodeHeader paramDef value =
+  let
+    encodedName =
+      Enc.encodeUtf8 (PD.parameterName paramDef)
+
+    encodedValue =
+      Enc.encodeUtf8 (PD.parameterRenderer paramDef value)
+  in
+    DList.singleton (CI.mk encodedName, encodedValue)
+
+{- |
+  An implementation of HeaderSchema for decoding queries
+-}
+newtype HeaderDecoder a b
+  = HeaderDecoder (HeaderMap -> Either T.Text b)
+
+decodeHeaders :: HeaderDecoder a b -> [HTTPTypes.Header] -> Either T.Text b
+decodeHeaders (HeaderDecoder parseMap) =
+  let
+    queryMap :: Ord k => [(k, a)] -> Map.Map k (DList.DList a)
+    queryMap =
+      Map.fromListWith (flip (<>))
+        . fmap (fmap DList.singleton)
+  in
+    parseMap . queryMap
+
+type HeaderMap = Map.Map (CI.CI BS.ByteString) (DList.DList BS.ByteString)
+
+instance PS.ParameterSchema HeaderDecoder where
+  newtype Parameter HeaderDecoder _record a = DecodeHeader (HeaderMap -> Either T.Text a)
+
+  makeParams constructor =
+    HeaderDecoder (\_headers -> Right constructor)
+
+  validateParams _unvalidate validate (HeaderDecoder f) =
+    HeaderDecoder (validate <=< f)
+
+  addParam (HeaderDecoder f) (DecodeHeader g) =
+    HeaderDecoder (\headers -> f headers <*> g headers)
+
+  required _accessor paramDef =
+    let
+      paramName =
+        PD.parameterName paramDef
+
+      paramNameBytes =
+        CI.mk
+          . Enc.encodeUtf8
+          $ paramName
+    in
+      DecodeHeader $ \headers -> do
+        mbValue <- PL.lookupSingleValue paramName paramNameBytes headers
+        case mbValue of
+          Just value -> PL.decodeParamBytes paramDef value
+          Nothing -> Left $ T.pack "Required header param missing: " <> paramName
+
+  optional _accessor paramDef =
+    let
+      paramName =
+        PD.parameterName paramDef
+
+      paramNameBytes =
+        CI.mk
+          . Enc.encodeUtf8
+          $ paramName
+    in
+      DecodeHeader $ \headers -> do
+        mbValue <- PL.lookupSingleValue paramName paramNameBytes headers
+        case mbValue of
+          Just value -> Just <$> PL.decodeParamBytes paramDef value
+          Nothing -> Right Nothing
+
+  splat _accessor (HeaderDecoder f) =
+    DecodeHeader f
+
+instance HeaderSchema HeaderDecoder where
+  type Cookies HeaderDecoder = CS.CookieDecoder
+
+  cookies _accessor cookieDecoder =
+    let
+      paramNameBytes =
+        HTTPTypes.hCookie
+
+      paramName =
+        -- NOTE: decodeUtf8 is unsafe in that it will throw an ugle impure
+        -- exception if the bytes are not valid utf8. That should never
+        -- happen with the Cookie header name though.
+        Enc.decodeUtf8 (CI.original paramNameBytes)
+    in
+      DecodeHeader $ \headers -> do
+        mbValue <- PL.lookupSingleValue paramName paramNameBytes headers
```

Comment:
In HTTP/2 there might be [more than one cookie header](https://www.rfc-editor.org/rfc/rfc9113#name-compressing-the-cookie-head)

Reply:
Interestingly, the language in the spec sort of implies that the request parsing should convert it to a single cookie before we see it. In any case, I'm going to defer this complication for now.